cpu/atmega_common/periph/spi.c

```diff
@@ -29,17 +29,56 @@
 #include "cpu.h"
 #include "mutex.h"
 #include "periph/spi.h"
+#include "macros/units.h"
+
+#define ENABLE_DEBUG        0
+#include "debug.h"
+
+static mutex_t lock = MUTEX_INIT;
 
 /**
  * @brief   Extract BR0, BR1 and SPI2X bits from speed value
  * @{
  */
-#define CLK_MASK            (0x3)
-#define S2X_SHIFT           (2)
+/* Helper function to compute a right shift value corresponding to
+ * dividers */
+static uint8_t _clk_shift(uint32_t clk)
+{
+    /* Atmega datasheets give the following table:
+     * SPI2X    SPR1    SPR0    SCK Frequency
+     * 0        0       0       Fosc/4
+     * 0        0       1       Fosc/16
+     * 0        1       0       Fosc/64
+     * 0        1       1       Fosc/128
+     * 1        0       0       Fosc/2
+     * 1        0       1       Fosc/8
+     * 1        1       0       Fosc/32
+     * 1        1       1       Fosc/64
+     *
+     * We can easily sort it by dividers by inverting the SPI2X bit and
+     * taking it as LSB:
+     * Divider  SPR1    SPR2    ~SPI2X
+     * 2        0       0       0
+     * 4        0       0       1
+     * 8        0       1       0
+     * 16       0       1       1
+     * 32       1       0       0
+     * 64       1       0       1
+     * 64       1       1       0
+     * 128      1       1       1
+     * Dividers are all power of two so we can apply them by
+     * right shifting these register value up to 5:
+     * clk = (CLOCK_CORECLOCK / 2) >> (SPR1 << 2 | SPR0 << 1 | ~SPI2X)
+     */
+    uint8_t shift = 0;
+    while (clk << shift < CLOCK_CORECLOCK / 2) {
+        shift++;
+    }
```

Comment:
```suggestion
    uint8_t shift = 0;
    uint32_t spi_clk = CLOCK_CORECLOCK / 2;
    while (spi_clk > clk) {
        shift++;
        spi_clk >>= 1;
    }
```

IMO this is easier to read, as the MCU will divide down the core clock to generate the SPI clock. Also, this will no longer overflow for high values of `clk`.